proposals/semi-auto-properties.md

```diff
@@ -10,6 +10,257 @@ Two common scenarios are that you want to apply a constraint on the setter, ensu
 
 In these cases by now you always have to create an instance field and write the whole property yourself.  This not only adds a fair amount of code, but it also leaks the `field` into the rest of the type's scope, when it is often desirable to only have it be available to the bodies of the accessors.
 
+## Detailed design
+
+For properties with an `init` accessor, everything that applies below to `set` would apply instead to the `init` accessor.
+
+**Principle 1:** Every property declaration can be thought of as having a backing field by default, which is elided when not used. The field is referenced using the keyword `field` and its visibility is scoped to the accessor bodies.
+
+**Principle 2:** `get;` will now be considered syntactic sugar for `get => field;`, and `set;` will now be considered syntactic sugar for `set => field = value;`.
+
+Both of these principles only apply under the same conditions where `{ get; }` or `{ get; set; }` already declares an auto property ([§15.7.4](https://github.com/dotnet/csharpstandard/blob/standard-v7/standard/classes.md#1574-automatically-implemented-properties)). Indexers also remain unaffected.
+
+This means that properties may now mix and match auto accessors with full accessors. For example:
+
+```cs
+{ get; set => Set(ref field, value); }
+```
+
+```cs
+{ get => field ?? parent.AmbientValue; set; }
+```
+
+Both accessors may be full accessors with either one or both making use of `field`:
+
+```cs
+{ get => field; set => field = value; }
+```
+
+```cs
+{ get => field; set => throw new InvalidOperationException(); }
+```
+
+```cs
+{ get => overriddenValue; set => field = value; }
+```
+
+```cs
+{
+    get;
+    set
+    {
+        if (field == value) return;
+        field = value;
+        OnXyzChanged();
+    }
+}
+```
+
+Expression-bodied properties and properties with only a `get` accessor may also use `field`:
+
+```cs
+public string LazilyComputed => field ??= Compute();
+```
+
+```cs
+public string LazilyComputed { get => field ??= Compute(); }
+```
+
+As with regular auto-properties, a setter that uses a backing field is disallowed when there is no getter. This restriction could be loosened in the future to allow the setter to do something only in response to changes, by comparing `value` to `field` (see open questions).
+
+```cs
+// ❌ Error, will not compile
+{ set => field = value; }
+```
+
+### Field-targeted attributes
+
+As with regular auto-properties, any property that uses a backing field in one of its accessors will be able to use field-targeted attributes:
```

Comment:
i both like that we're claling this out, but would like a different way of phrasing it.  Basically, i would say that `[field: ...]` can be used on an auto prop.  And as these new properties are auto-props that continues to apply.

Reply:
I don't like that we would call these auto-props, actually. The principle is that every field has a backing field, elided if unused. `{ get => field; set => field = value; }` is not an auto-prop or else every property is an auto prop because it has `field` available (elided if unused).

Thus my strong resistance to titling the feature "semi-auto props" and my relief that it switched to "`field` keyword"

Reply:
I think the concept of "it always exists but might be elided" doesn't quite track for me. It perhaps depends at which stage is the thing elided.
- Is it elided for purposes of struct circularity checks? It must be, otherwise `struct S { S Prop { get => ...; } }` is broken, but that does work today.
- Is it elided for purposes of public API? e.g. if I ask the property symbol for the "AssociatedField" when the field is not used, do I get a symbol back? (I might have the details wrong, but this still seems relevant, I think the backing fields are exposed through the public APIs.)

Basically my expectation is that we introduced `field` as a contextual keyword in accessors and this gives us the ability to decide if the field is being used with a pure syntactic check and not actually binding anything. Does the model of "not using `field` causes it to be elided" have a difference in practice from "an implicit field is added to containing type when `field` is used"?

Reply:
it's 'elided' conceptually at the lang level.  Meaning it simply does not exist.  As such, for things like "Struct circularity" since it does not exist it does not effect things.


Elided just means 'does not exist', and everything else should flow from that.  It's not 'it exist,s but it not emitted'.  IMO.

Reply:
And also, I should have said, even apart from the 'elide' concept, I still don't think `{ get => field; set => field = value; }` is in the spirit of what the spec meant by "automatically implemented" just because the backing field can be referred to by a keyword instead of manually _declared_.

Reply:
Currently, "is it an auto property" is conflated with the question "does it have an implicit backing field". I am open to these concepts being separated as part of this proposal, but we'll have to be careful to thread that change all the way through the spec and the implementation.

Reply:
I added a clarification to that effect.